### PR TITLE
Runtime performance improvement with sampling

### PR DESCRIPTION
**Pull request by JingJZ160**
### Description
Adding sampling method in the data cleaning and intent resolving steps. No performance testing has done yet. This is a placeholder but please check the style, logic and whether we need another API to override sampling percentage.

### CODE REVIEW

File: foreshadow/foreshadow.py [hunk 1]
```diff
@@ -555,3 +554,32 @@ def configure_multiprocessing(self, n_job: int = 1) -> NoReturn:
...
+        Returns:
+
+        """
+        self.X_preparer.cache_manager["config"][
+            ConfigKey.ENABLE_SAMPLING
+        ] = enable_sampling
+        self.X_preparer.cache_manager["config"][
+            ConfigKey.SAMPLING_FRACTION
+        ] = sampling_fraction
+        self.X_preparer.cache_manager["config"][
```
Comment (jichaogp) [inline foreshadow/foreshadow.py]: Better to make 'config' a constant as well.

File: foreshadow/tests/test_core/test_cache_manager.py [hunk 2]
```diff
@@ -236,7 +236,11 @@ def test_cache_manager_dict_serialize(store):
         else:
             expected["store"][key] = PrettyDefaultDict(lambda: None)
 
-    assert expected == cs.dict_serialize(deep=True)
+    # TODO it's an ugly fix. Idealy we should add the default config section
+    #  back but it will make the json file really bulky.
+    cs_serialized = cs.dict_serialize(deep=True)
+    cs_serialized["store"].pop("config")
+    assert expected == cs_serialized
```
Comment (jichaogp) [inline foreshadow/tests/test_core/test_cache_manager.py]: Please also check the config object popped from the dictionary.

File: foreshadow/tests/test_foreshadow.py [hunk 3]
```diff
@@ -1050,3 +1050,36 @@ def test_foreshadow_serialization_tpot():
     # converge. The test here aims to evaluate if both cases have
     # produced a reasonable score and the difference is small.
     assert score1 > 0.9 and score2 > 0.9
+
+
+def test_foreshadow_configure_sampling():
```
Comment (jichaogp) [inline foreshadow/tests/test_foreshadow.py]: for this test case, maybe copy an existing integration test case but enable sampling to verify the sampling logic is working.

File: foreshadow/foreshadow.py (hunk 1 above)
Comment (JingJZ160) [reply foreshadow/foreshadow.py]: Done.

File: foreshadow/tests/test_core/test_cache_manager.py (hunk 2 above)
Comment (JingJZ160) [reply foreshadow/tests/test_core/test_cache_manager.py]: Done.

File: foreshadow/tests/test_foreshadow.py (hunk 3 above)
Comment (JingJZ160) [reply foreshadow/tests/test_foreshadow.py]: Done.

**Pull request merged**